enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,164 @@
+currently is where crc, oc, ocp-dev-preview, ocp, odo, and serverless artifacts are published.  This is currently how OpenShift CLIs are published.  The ConsoleCLIDownloads
+CRs reference this index.  The challenge is offering these in disconnected environments and automating the download and extraction of the artifacts.  
+
+3.  Images for each CLI. The CLI in-cluster manager would provide an extract mechanism for each CLI image, similar to how we currently 
+`oc adm release extract --command oc` and `oc image extract`.   A Custom Resource for each CLI would provide the in-cluster CLI manager information about each CLI and its image.
+In disconnected, the images will be available through a mirrored local registry.  The logic for `oc image extract` can be re-used and extended for extracting the CLIs.  CLIs would
+be managed by users through extending the oc commands we currently have to extract `oc` and `openshift-install` binaries from the release payload.  This would provide the function we need, but looking to the future there is an
+upstream effort we can use to provide an in-cluster CLI manager while also providing a mechanism that the Kubernetes community as a whole can utilize.  That brings us to option 4.
+
+4.  The CLI in-cluster manager can provide CLIs as plugins through [krew](https://github.com/kubernetes-sigs/krew).  CLI images will serve the artifacts to enable disconnected downloads, and OpenShift CLIs will be available as 
+[krew plugins](https://github.com/kubernetes-sigs/krew-index/tree/master/plugins).
+    * Krew and Krew plugins are upstream projects that Kubernetes users are already familiar with
+    * From the [README](https://github.com/kubernetes-sigs/krew/#what-does-krew-do): `Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is similar to tools like apt, dnf or brew. Today, over 70 kubectl plugins are available on Krew.` It makes sense to offer OpenShift CLIs in the same way.
+    * With krew, artifacts will be packaged within an image.  Each cluster will have an index of CLI images.  Mirrored local registries will provide binaries to disconnected environments.  
+
+Option 4 is the option we as a team are leaning toward.  It requires that each OpenShift CLI publish artifacts, for example, [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+then create an image similar to this: [example cli-artifacts image](https://github.com/openshift/oc/blob/master/images/cli-artifacts/Dockerfile.rhel).  CLI artifacts are packaged in images.
+The CLI in-cluster manager will gather information from each CLI's CR about each CLI and its image.  A Krew index will be populated in every cluster to provide available CLIs and plugins for a cluster.
+
+### Non-Goals
+
+* CLI manager will not build or serve the binaries.  It will know where to find them.
```

Comment:
Seems like making each CLI providing come up with a way to service each binary provides overhead. Could it not just place binaries somewhere (or update reference) and have a common pod/method to serve the binaries? At very least, it would be good to provide a reusable framework for each CLI provider to use to wrap/serve their binary. Disconnected is an important part as well, so providing guidance on how binaries are distributed in a consistent way I would think would help most.

Reply:
The binaries are meant to be served from images, which are already built. We're not aspiring to coming up with a new distribution mechanism.